plugin-management-library/src/main/java/io/jenkins/tools/pluginmanager/impl/PluginManager.java

```diff
@@ -0,0 +1,546 @@
+            }
+        }
+
+        createLocks(plugins);
+
+        List<String> bundledPlugins = bundledPlugins();
+        List<String> installedPlugins = installedPlugins();
+
+        downloadPlugins(plugins);
+
+        writeFailedPluginsToFile();
+
+        //clean up locks
+
+    }
+
+
+    public void getSecurityWarnings() {
+        JSONObject updateCenterJson = getUpdateCenterJson();
+        JSONArray warnings = updateCenterJson.getJSONArray("warnings");
+
+        for (int i = 0; i < warnings.length(); i++) {
+            JSONObject warning = warnings.getJSONObject(i);
+            String warningType = warning.getString("type");
+            if (!warningType.equals("plugin")) {
+                continue;
+            }
+            String warningId = warning.getString("id");
+            String warningMessage = warning.getString("message");
+            String warningName = warning.getString("name");
+            String warningUrl = warning.getString("url");
+
+            SecurityWarning securityWarning = new SecurityWarning(warningId, warningMessage, warningName, warningUrl);
+            JSONArray warningVersions = warning.getJSONArray("versions");
+            for (int j = 0; j < warningVersions.length(); j++) {
+                JSONObject warningVersion = warningVersions.getJSONObject(j);
+                String lastVersion = "";
+                if (warningVersion.has("lastVersion")) {
+                    lastVersion = warningVersion.getString("lastVersion");
+                }
+                String pattern = warningVersion.getString("pattern");
+                securityWarning.addSecurityVersion(lastVersion, pattern);
+            }
+            allSecurityWarnings.add(securityWarning);
+        }
+    }
+
+
+    public void checkVersionSpecificUpdateCenter() {
+        //check if version specific update center
+        if (!StringUtils.isEmpty(jenkinsVersion)) {
+            JENKINS_UC_LATEST = new StringBuilder(JENKINS_UC).append(jenkinsVersion).toString();
+
+            try (CloseableHttpClient httpclient = HttpClients.createDefault();) {
+                HttpGet httpget = new HttpGet(JENKINS_UC_LATEST);
+                try (CloseableHttpResponse response = httpclient.execute(httpget);) {
+                    if (response.getStatusLine().getStatusCode() != HttpStatus.SC_OK) {
+                        JENKINS_UC_LATEST = "";
+                    }
+                } catch (IOException e) {
+                    JENKINS_UC_LATEST = "";
+                    System.out.println("No version specific update center for Jenkins version " + jenkinsVersion);
+                }
+            } catch (IOException e) {
+                JENKINS_UC_LATEST = "";
+                System.out.println("Unable to check if version specific update center for Jenkins version " + jenkinsVersion);
+            }
+
+        }
+
+    }
+
+
+    public void writeFailedPluginsToFile() {
+        FileWriter fileWriter;
+
+        try {
```

Comment:
FileWriter also implements auto closeable so you can use try with resources:
https://docs.oracle.com/javase/8/docs/api/index.html?java/io/FileWriter.html
Also in your current implementation if there's an exception you aren't closing the file
```suggestion
        try (FileWriter fileWriter = new FileWriter("failedplugins.txt")) {
```

Reply:
try, catch, finally.  Though +1 on filewriter if it's auto-closing everything.